spyder_kernels/customize/namespace_manager.py

```diff
@@ -39,49 +37,52 @@ class NamespaceManager:
     current_namespace is True, or a new namespace.
     """
 
-    def __init__(self, filename, namespace=None, current_namespace=False,
-                 file_code=None, stack_depth=1):
+    def __init__(
+        self,
+        shell,
+        filename,
+        current_namespace=False,
+        file_code=None,
+        context_locals=None,
+        context_globals=None,
+    ):
+        self.shell = shell
         self.filename = filename
-        self.ns_globals = namespace
+        self.ns_globals = None
         self.ns_locals = None
         self.current_namespace = current_namespace
         self._previous_filename = None
         self._previous_main = None
-        self._previous_running_namespace = None
         self._reset_main = False
         self._file_code = file_code
-        ipython_shell = get_ipython()
-        self.context_globals = ipython_shell.get_global_scope(stack_depth + 1)
```

Comment:
stack depth manipulations are very hacky. We assume that this is run in the user_ns which will work in all valid cases (i.e. to break this you would need to add runfile in a library and not specify the namespace it should be run in)